clang/lib/Sema/SemaTemplate.cpp

```diff
@@ -6930,17 +6934,20 @@ ExprResult Sema::CheckTemplateArgument(NonTypeTemplateParmDecl *Param,
     IsConvertedConstantExpression = false;
   }
 
-  if (getLangOpts().CPlusPlus17) {
+  if (getLangOpts().CPlusPlus17 || PartialOrderingTTP) {
     // C++17 [temp.arg.nontype]p1:
     //   A template-argument for a non-type template parameter shall be
     //   a converted constant expression of the type of the template-parameter.
     APValue Value;
     ExprResult ArgResult;
     if (IsConvertedConstantExpression) {
-      ArgResult = BuildConvertedConstantExpression(Arg, ParamType,
-                                                   CCEK_TemplateArg, Param);
-      if (ArgResult.isInvalid())
+      ArgResult = BuildConvertedConstantExpression(
+          Arg, ParamType,
+          PartialOrderingTTP ? CCEK_InjectedTTP : CCEK_TemplateArg, Param);
+      if (ArgResult.isInvalid()) {
```

Comment:
`!isUsable` is perhaps more useful here (since it is invalid + not initialized)?

Reply:
I don't think any of the Build*/Act* functions return uninitialized ExprResult. I see that it's rarely used at all, some functions return it to mean "some unknown expression" due to dependence, but that's happening more in other parts of the code base, like the concepts implementation.

Maybe we should split these users of ExprResult into ExprResultOrNone, and make ExprResult always initialized.

Reply:
ExprResult always being initialized isn't a bad idea... but a much bigger change than i'd request here.  That said, the `Build/Act` functions SOMETIMES return uninit (even if by accident sometimes, particularly when used as an out param).  So the `isUsable` is both what we typically mean, and the safer option.

I realize we are bad about `isInvalid` all over the place, I've actively been trying to stop using it as we never MEAN it, so I show a strict preference for `isUsable`, particularly in these cases.  IMO, `isInvalid` becoming deprecated is perhaps of value.

Reply:
The big question for me about updating these users around here would be: Does the function returning uninitialized mean that's an error, and has that error already been diagnosed?

Reply:
In every singe case I've seen this come up, it has meant either 'error' or 'there is nothing here'.  And the question we usually ask is, "Can I do something with this".  Which `isUsable` means.  `isInvalid` means, "WAS something here, but there was a problem with it".

In this case, you're translating "There is nothing I can do with this" to "error".  Which makes more sense as `isUsable` to me.

Reply:
In this case, I want : "has this function diagnosed an error?" if so, I want to attach a note to that error, pointing to the relevant template parameter.

Reply:
In that case, what do you wish to do if the function has NOT diagnosed an error, but has returned nothing?  are you properly handling that here?

Reply:
The problem is that I don't know what uninitialized would mean here, this is pre-existing.
Would I want to attach the note anyway? Would I want to forward that uninitialized to the caller, instead of turning it into an ExprError?

I put an `assert(!ArgResult.isUnset());` here, it doesn't trip a single test case.

Reply:
I know it is not representative of this test case not existing, but uninitialized return is not documented here, and I couldn't find such a case by inspecting the code.

I would rather assert than guess what should happen here, thanks!

Reply:
>I know it is not representative of this test case not existing, but uninitialized return is not documented here, and I couldn't find such a case by inspecting the code.

THAT is precisely the concern.